src/Refinery/ProofState.hs

```diff
@@ -32,23 +34,48 @@ import           Control.Monad.Except
 import qualified Control.Monad.Writer.Lazy as LW
 import qualified Control.Monad.Writer.Strict as SW
 import           Control.Monad.State
-import           Control.Monad.Logic
 import           Control.Monad.Morph
 import           Control.Monad.Reader
-import           Data.Either
 
 import           GHC.Generics
 
+-- | The core data type of the library.
+-- This represents a reified, in-progress proof strategy for a particular goal.
+--
+-- NOTE: We need to split up the extract type into @ext'@ and @ext@, as
+-- we want to be functorial (and monadic) in @ext@, but it shows up in both
+-- co and contravariant positions. Splitting the type variable up into two solves this problem,
+-- at the cost of looking ugly.
 data ProofStateT ext' ext err s m goal
     = Subgoal goal (ext' -> ProofStateT ext' ext err s m goal)
+    -- ^ Represents a subgoal, and a continuation that tells
+    -- us what to do with the resulting extract.
     | Effect (m (ProofStateT ext' ext err s m goal))
+    -- ^ Run an effect.
     | Stateful (s -> (s, ProofStateT ext' ext err s m goal))
+    -- ^ Run a stateful computation. We don't want to use 'StateT' here, as it
+    -- doesn't play nice with backtracking.
     | Alt (ProofStateT ext' ext err s m goal) (ProofStateT ext' ext err s m goal)
+    -- ^ Combine together the results of two @ProofState@s.
```

Comment:
Is there any reason to have alt?

Reply:
Yeah, it has to do with interleaving of results + the identity laws.